pkg/gpu/stream.go

```diff
@@ -162,37 +183,79 @@ func (sh *StreamHandler) handleKernelLaunch(event *gpuebpf.CudaKernelLaunch) {
-		if logLimitErrorAttach.ShouldLog() {
+		if logLimitStreams.ShouldLog() {
 			log.Warnf("Error attaching kernel data for PID %d: %v", sh.metadata.pid, err)
 		}
 	}
 
 	sh.kernelLaunches = append(sh.kernelLaunches, *enrichedLaunch)
+
+	// If we've reached the kernel launch limit, trigger a sync. This stops us from just collecting
+	// kernel launches and not generating any spans if for some reason we are missing sync events.
+	if len(sh.kernelLaunches) >= sh.limits.maxKernelLaunches {
+		sh.markSynchronization(event.Header.Ktime_ns + 1) // sync "happens" after the launch, not the same time. If the time is the same, the last kernel launch is not included in the span
+		sh.telemetry.forcedSyncOnKernelLaunch.Inc()
+	}
 }
 
 func (sh *StreamHandler) handleMemEvent(event *gpuebpf.CudaMemEvent) {
 	if event.Type == gpuebpf.CudaMemAlloc {
-		sh.memAllocEvents[event.Addr] = *event
+		_ = sh.memAllocEvents.Add(event.Addr, *event)
 		return
 	}
 
 	// We only support alloc and free events for now, so if it's not alloc it's free.
-	alloc, ok := sh.memAllocEvents[event.Addr]
+	alloc, ok := sh.memAllocEvents.Get(event.Addr)
 	if !ok {
-		log.Warnf("Invalid free event: %v", event)
+		if logLimitStreams.ShouldLog() {
+			log.Warnf("Invalid free event: %v", event)
+		}
+		sh.telemetry.invalidFreeEvents.Inc()
 		return
 	}
 
+	sh.createMemoryAllocation(event.Header.Ktime_ns, &alloc, false)
+
+	sh.suspendEvictCallback = true
+	defer func() {
+		sh.suspendEvictCallback = false
+	}()
+	sh.memAllocEvents.Remove(event.Addr)
+}
+
+func (sh *StreamHandler) createMemoryAllocation(freeKtimeNs uint64, alloc *gpuebpf.CudaMemEvent, isLeaked bool) {
 	data := memoryAllocation{
 		startKtime: alloc.Header.Ktime_ns,
-		endKtime:   event.Header.Ktime_ns,
+		endKtime:   freeKtimeNs,
 		size:       alloc.Size,
 		allocType:  globalMemAlloc,
-		isLeaked:   false, // Came from a free event, so it's not a leak
+		isLeaked:   isLeaked,
 	}
 
 	sh.allocations = append(sh.allocations, &data)
-	delete(sh.memAllocEvents, event.Addr)
+}
+
+func (sh *StreamHandler) onEvictedAlloc(_ uint64, value gpuebpf.CudaMemEvent) {
+	// If we are suspending the evict callback, don't create memory allocations.
+	// The reason is that the LRU cache calls the eviction callback whenever an item is removed for
+	// any cause (full cache or manual removal). In handleMemEvent, we manually remove items from the cache,
+	// so we don't want to create another memory allocation in that case.
+	if sh.suspendEvictCallback {
+		return
+	}
+
+	nowTs, err := ddebpf.NowNanoseconds()
+	if err != nil {
+		if logLimitStreams.ShouldLog() {
+			log.Warnf("Error getting current time: %v", err)
+		}
+		return
+	}
+
+	// This is probably a leak, it's an allocation that was never freed and it got to the point
+	// where our cache was full. But we don't know for sure, so don't mark it as a leak.
+	sh.createMemoryAllocation(uint64(nowTs), &value, false)
```

Comment:
that is also very confusing and feels like it should be here.
evict handler should have only one job: clean the data structure and not create additional entries

Reply:
lets keep it as simple as possible. let's loose the data (maybe "log" it via some internal telemetry to track the frequency of this edge-case